### PR TITLE
[prometheusexporter] Expose job and instance labels

### LINKED ISSUE #8355: [prometheusexporter] Add job and instance labels to match federated prometheus
**Description**
As described in the [OpenTelemetry -> prometheus conversion specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification%2Fmetrics%2Fdatamodel.md#resource-attributes-1), prometheus exporters must add job and instance labels to exported metrics in the collector:

> In the collector Prometheus exporters, the service.name and service.namespace attributes MUST be combined as <service.namespace>/<service.name>, or <service.name> if namespace is empty, to form the job metric label. The service.instance.id attribute, if present, MUST be converted to the instance label; otherwise, instance should be added with an empty value.

This allows a user to scrape the prometheuxexporter's endpoint with `honor_labels: true` to preserve the original job and instance labels as described in https://prometheus.io/docs/prometheus/latest/federation/#configuring-federation.

Currently, the prometheus exporter only adds metric attributes: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5cd8dd90659779c0d6f99873de71144a2d5235d1/exporter/prometheusexporter/collector.go#L82-L98

### PR DESCRIPTION
**Pull request by gouthamve**
**NOTE**: This is WIP PR and I wanted to understand if my approach is right before I write the tests for this.

**Description:**

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8355

**Testing:** I've run the collector with the following config and the exposed metrics had `job` and `instance` attached to them:

```
receivers:
  prometheus:
    config:
      scrape_configs:
      - job_name: 'beep-boop'
        scrape_interval: 5s
        static_configs:
        - targets: ['localhost:9090']

processors:
  batch:

exporters:
  logging:
    logLevel: info

  prometheus:
    endpoint: localhost:9091

service:
  pipelines:
    metrics:
      receivers: [prometheus]
      processors: [batch]
      exporters: [prometheus]
```

### CODE REVIEW

**Comment by gouthamve** [comment]
@dashpole Could you look at my approach of plumbing the resource attributes all the way down, and let me know if it looks okay? If it does, I will work on adding tests.

**Comment by dashpole** [comment]
It looks good to me!

**Comment by gouthamve** [comment]
I've added the tests and the tests caught a bug 😀 @dashpole I think this is ready for review!

File: exporter/prometheusexporter/collector.go [hunk 1]
```diff
@@ -79,16 +81,41 @@ func (c *collector) metricName(namespace string, metric pdata.Metric) string {
...
+	values := make([]string, 0, attributes.Len()+2)
 
 	attributes.Range(func(k string, v pdata.Value) bool {
 		keys = append(keys, sanitize(k, c.skipSanitizeLabel))
 		values = append(values, v.AsString())
 		return true
 	})
 
+	// Sometimes the map can be empty and lookups on an empty map panic.
+	if (resourceAttrs == pdata.Map{}) {
```
Comment (dashpole) [inline exporter/prometheusexporter/collector.go:95]: I don't think we need to handle this case.  We don't do this for metric attributes, for example (we Range on it above).  Tests should use pdata.NewMap().

Comment (gouthamve) [reply exporter/prometheusexporter/collector.go]: Good catch, removing this check surfaced another bug 🙃 

I think it might be ready to go now!

**Review by dashpole — APPROVED** [review]

**Comment by Mario-Hofstaetter** [comment]
@gouthamve Could you please update https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/README.md to explain the behavior regarding new `job` and `instance` labels?
Is `job_name` from `scrape_configs` used as `job` ?

I have added those labels myself in a way so have to be careful not to break anything.

**Comment by gouthamve** [comment]
> Is job_name from scrape_configs used as job ?

Yes. I'll update the docs now.

**Pull request merged**